examples/java8/src/main/java/org/apache/beam/examples/advanced/subprocess/utils/CallingSubProcessUtils.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.examples.advanced.subprocess.utils;
+
+import com.google.common.collect.Sets;
+
+import java.util.Set;
+import java.util.concurrent.Semaphore;
+
+import org.apache.beam.examples.advanced.subprocess.configuration.SubProcessConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility class for dealing with concurrency and binary file copies to the worker.
+ */
+public class CallingSubProcessUtils {
+
+  // Prevent Instantiation
+  private CallingSubProcessUtils() {};
+
+  static final Logger LOG = LoggerFactory.getLogger(CallingSubProcessUtils.class);
+
+  static boolean initCompleted = false;
+
+  // Allow multiple subclasses to create files, but only one thread per subclass can add the file to
+  // the worker
+  private static final Set<String> downloadedFiles = Sets.<String>newConcurrentHashSet();
+
+  // Limit the number of threads able to do work
+  private static Semaphore semaphore;
+
+  public static void setUp(SubProcessConfiguration configuration, String binaryName)
+      throws Exception {
+
+    if (!initCompleted) {
+
+      if (semaphore == null) {
+        initSemaphore(configuration.getConcurrency());
+      }
+
+      synchronized (downloadedFiles) {
+        if (!downloadedFiles.contains(binaryName)) {
+          // Create Directories if needed
+          FileUtils.createDirectoriesOnWorker(configuration);
+          LOG.info("Calling filesetup to move Executables to worker.");
+          ExecutableFile executableFile = new ExecutableFile(configuration, binaryName);
+          FileUtils.copyFileFromGCSToWorker(executableFile);
```

Comment:
If the process restarts, the file might be in the worker but not in downloadedFiles. Will this copy throw an exception in that case?

Reply:
It should just over write the file, if I do check for existence before the move to avoid a unneeded copy I would also need to confirm the checksum. Happy to add that in if you think its more useful.